peer_test.go

```diff
@@ -217,3 +217,122 @@ func TestPeerSelectionPreferIncoming(t *testing.T) {
 		c.Close()
 	}
 }
+
+type peerTest struct {
+	t        testing.TB
+	channels []*Channel
+}
+
+// NewService will return a new server channel and the host port.
+func (pt *peerTest) NewService(t testing.TB, svcName string) (*Channel, string) {
+	ch := testutils.NewServer(t, &testutils.ChannelOpts{ServiceName: svcName})
+	pt.channels = append(pt.channels, ch)
+	return ch, ch.PeerInfo().HostPort
+}
+
+// CleanUp will clean up all channels started as part of the peer test.
+func (pt *peerTest) CleanUp() {
+	for _, ch := range pt.channels {
+		ch.Close()
+	}
+}
+
+type peerSelectionTest struct {
+	peerTest
+
+	// numPeers is the number of peers added to the client channel.
+	numPeers int
+	// numAffinity is the number of affinity nodes.
+	numAffinity int
+	// numConcurrent is the number of concurrent goroutine to make outbound calls.
+	numConcurrent int
+	// hasInboundCall is the bool flag to tell whether to have inbound calls from affinity nodes
+	hasInboundCall bool
+
+	servers  []*Channel
+	affinity []*Channel
+	client   *Channel
+}
+
+// setupServers will create numPeer servers, and register handlers on them.
+func (pt *peerSelectionTest) setupServers(t testing.TB) {
+	pt.servers = make([]*Channel, pt.numPeers)
+
+	// Set up numPeers servers.
+	for i := 0; i < pt.numPeers; i++ {
+		pt.servers[i], _ = pt.NewService(t, "hyperbahn")
+		pt.servers[i].Register(raw.Wrap(newTestHandler(pt.t)), "echo")
+	}
+}
+
+func (pt *peerSelectionTest) setupAffinity(t testing.TB) {
+	pt.affinity = make([]*Channel, pt.numAffinity)
+	for i := range pt.affinity {
+		pt.affinity[i] = pt.servers[i]
+	}
+
+	hostport := pt.client.PeerInfo().HostPort
+	serviceName := pt.client.PeerInfo().ServiceName
+	var wg sync.WaitGroup
+	wg.Add(pt.numAffinity)
+	// Connect from the affinity nodes to the service.
+	for _, affinity := range pt.affinity {
+		affinity.Peers().Add(hostport)
```

Comment:
It is the same, which is why you might as well put it in the goroutine. Then the user's context when reading this code is: each loop starts a goroutine for each affinity node, and that affinity node does <blah> work.

Currently, the user's context is:
each loop adds a peer to the affinity, *then* starts a goroutine. Usually that pattern means that something had to be done outside of the goroutine (e.g. for sync reasons), which is not the case here.

It just helps break apart the problem into simpler logical parts -- and in this case, reduces the number or "parts" you have to think about.